src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -4996,9 +4996,20 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           typedSelect(tree, typedSelectOrSuperQualifier(qual), nme.CONSTRUCTOR)
         case Select(qual, name) =>
           if (Statistics.canEnable) Statistics.incCounter(typedSelectCount)
-          val qualTyped = checkDead(typedQualifier(qual, mode))
+          def typeQual = checkDead(typedQualifier(qual, mode))
+          val (qualTyped, isJavaTypeSelection) = {
+            if (name.isTermName) (typeQual, false)
+            else {
+              qual match {
+                case id: Ident if context.unit.isJava =>
+                  val qualTyped1 = context.withinJavaTypeSelection(typeQual)
+                  (qualTyped1, qualTyped1.isType)
```

Comment:
It's really interesting that this is all we need to do in `typedSelectOrSuperCall`, since the whole method will return `Select(qualTyped1, name)` whereas the more correct tree would have been `SelectFromTypeTree(qualTyped1, name)`. I have not done this change because of simplicity. But it does make me wonder what is the crucial difference between `Select` and `SelectFromTypeTree`... It seems that `Select(Ident(TypeName("Foo")), _) == SelectFromTypeTree(Ident(TypeName("Foo")), _)`, no?

Reply:
I think for consistency it would be better to generate the `SelectFromTypeTree`, even if a `Select` doesn't violate any of the current assumptions in the rest of the compiler.

Reply:
I can do that, but I'm afraid to modify too much of the logic of this method since it's really hot. I also think that for consistency it's better to generate a `SelectFromTypeTree`. Doing that requires me to wrap almost all the rest of the body that follows this statement with an `if`.

Reply:
This should also get a comment that explains why `context.unit.isJava` here needs to be reified as a `context.withinJavaTypeSelection` for consumption down the line in `typedIdent`.